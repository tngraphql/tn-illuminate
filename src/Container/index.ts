/**
 * (c) Phan Trung NguyÃªn <nguyenpl117@gmail.com>
 * User: nguyenpl117
 * Date: 3/11/2020
 * Time: 12:32 PM
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

export { Container } from './Container';
export { Injector } from './Injector';
export { InvalidArgumentException } from './InvalidArgumentException';
export { LogicException } from './LogicException';
export { BindingResolutionException } from './BindingResolutionException';