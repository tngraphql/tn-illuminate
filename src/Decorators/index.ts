/**
 * (c) Phan Trung NguyÃªn <nguyenpl117@gmail.com>
 * User: nguyenpl117
 * Date: 3/21/2020
 * Time: 4:51 PM
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

export * from './Inject';
export * from './Rules';
export * from './Service';
export * from './UseMiddleware';
export * from './ValidateArgs';