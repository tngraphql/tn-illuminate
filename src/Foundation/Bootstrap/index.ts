/**
 * (c) Phan Trung NguyÃªn <nguyenpl117@gmail.com>
 * User: nguyenpl117
 * Date: 3/11/2020
 * Time: 12:34 PM
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

export { BootProviders } from './BootProviders';
export { LoadConfiguration } from './LoadConfiguration';
export { LoadEnvironmentVariables } from './LoadEnvironmentVariables';
export { RegisterProviders } from './RegisterProviders';
