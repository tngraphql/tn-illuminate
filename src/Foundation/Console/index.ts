/**
 * (c) Phan Trung NguyÃªn <nguyenpl117@gmail.com>
 * User: nguyenpl117
 * Date: 3/14/2020
 * Time: 10:26 AM
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

export * from './ConsoleKernel';
export * from './AceApplication';
export * from './CommandMakeCommand';
export * from './GeneratorCommand';
export * from './KeyGenerateCommand';
export * from './MiddlewareMakeCommand';
export * from './MigrateMakeCommand';
export * from './ModelMakeCommand';
export * from './ProviderMakeCommand';
export * from './ResolveMakeCommand';
export * from './TypeMakeCommand';
