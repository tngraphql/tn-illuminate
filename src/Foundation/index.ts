/**
 * (c) Phan Trung NguyÃªn <nguyenpl117@gmail.com>
 * User: nguyenpl117
 * Date: 3/11/2020
 * Time: 12:35 PM
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

export { BootProviders, LoadConfiguration, LoadEnvironmentVariables, RegisterProviders } from './Bootstrap';
export { Kernel } from './Kernel';
export { Application } from './Application';
export * from './Console';
export * from './GraphQL';