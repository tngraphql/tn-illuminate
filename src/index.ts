/**
 * (c) Phan Trung NguyÃªn <nguyenpl117@gmail.com>
 * User: nguyenpl117
 * Date: 3/11/2020
 * Time: 12:31 PM
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

export { Repository } from './config/Repository';
export { Container, Injector, InvalidArgumentException, BindingResolutionException, LogicException } from './Container';
export *  from './Decorators';
export {
    Application, ConsoleKernel, BootProviders, RegisterProviders, LoadEnvironmentVariables, LoadConfiguration
} from './Foundation';
export { ServiceProvider } from './Support/ServiceProvider';
export { ensureIsFunction, isClass, isFuntion, isEsm, isObject } from './utils';
export * from '@tngraphql/console';
