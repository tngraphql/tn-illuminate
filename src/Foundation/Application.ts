/**
 * Created by Phan Trung NguyÃªn.
 * User: nguyenpl117
 * Date: 3/9/2020
 * Time: 9:27 PM
 */
import { Container } from '../Container/Container';
import { esmRequire, resolveFrom } from '@poppinss/utils/build';
import { ServiceProvider } from '../Support/ServiceProvider';
import * as path from 'path';
import { EventServiceProvider } from './Events/EventServiceProvider';
import { ServiceProviderContract } from '../Contracts/ServiceProviderContract';
import { RoutingServiceProvider } from './Routing/RoutingServiceProvider';
import { ProviderRepository } from './ProviderRepository';
import { Filesystem } from '@poppinss/dev-utils/build';
import { Env } from '../Support/Env';
import { ApplicationContract } from '../Contracts/ApplicationContract';
import { LogServiceProvider } from '../Log/LogServiceProvider';
import { ProfilerServiceProvider } from '../Profiler/ProfilerServiceProvider';
import { LocaleUpdated } from './Events/LocaleUpdated';
import { event } from '../Support/helpers';

export class Application extends Container implements ApplicationContract{
    [key: string]: any;

    public get inProduction() {
        return process.env.NODE_ENV === 'production';
    };

    protected _basePath: string = process.cwd();

    protected serviceProviders: any[] = [];

    protected _booted: boolean = false;

    protected _bootedCallbacks: any[] = [];

    protected _instances = new Map();

    constructor(basePath: string = null) {
        super();

        if ( basePath ) {
            this.setBasePath(basePath);
        }

        this.registerBaseBindings();
        this.registerBaseServiceProviders();
    }

    public setBasePath(basePath: string) {
        this._basePath = basePath;
    }

    public basePath(p: string = ''): string {
        return path.join(this._basePath, p);
    }

    public getBasePath() {
        return this._basePath;
    }

    public configPath() {
        return this._basePath + '/config';
    }

    public instance(name: string, instance: any) {
        this[name] = instance;
    }

    public getter(name, callback) {
        (this as any).constructor.getter(name, callback);
    }

    protected registerBaseBindings() {
        Container['setInstance'](this);

        this.bind(Container, () => this);
        this.bind(Application, () => this);
    }

    protected registerBaseServiceProviders() {
        try {
            this.register(new EventServiceProvider(this));
            this.register(new ProfilerServiceProvider(this));
            this.register(new LogServiceProvider(this));
            this.register(new RoutingServiceProvider(this));
        } catch (e) {
            throw e;
        }
    }

    public getServiceProviders() {
        return this.serviceProviders;
    }

    public async registerConfiguredProviders() {
        await new ProviderRepository(this as any, new Filesystem(this.basePath()), this.getCachedServicesPath()).load(this.config.get('app.providers', []));

        /*this.config.get('app.providers', []).map(async provider => {
            if ( typeof provider === 'string' ) {
                await this.register(this.resolveProvider(provider))
            } else {
                await this.register(new provider(this));
            }
        });*/
    }

    public async bootstrapWith(bootstrappers: any[]) {
        for await(const bootstrapper of bootstrappers ) {
            await this.make<any>(bootstrapper).bootstrap(this);
        }
    }

    /**
     * Register a service provider with the application.
     *
     * @param provider
     * @param force
     */
    public async register(provider: ServiceProviderContract | string, force: boolean = false) {

        if ( typeof provider === 'string' ) {
            provider = await this.resolveProvider(provider);
        }

        if ( typeof ((provider as any).register) === 'function' ) {
            await (provider as any).register();
        }

        if ( (provider as object).hasOwnProperty('bindings') ) {
            for( let item of provider['bindings'] ) {
                if ( ! Array.isArray(item) ) {
                    item = [item];
                }
                const [namespace, concrete] = item;
                this.bind(namespace, concrete || namespace);
            }
        }

        if ( (provider as object).hasOwnProperty('singletons') ) {
            for( let item of provider['singletons'] ) {
                if ( ! Array.isArray(item) ) {
                    item = [item];
                }
                const [namespace, concrete] = item;
                this.singleton(namespace, concrete || namespace);
            }
        }

        await this.markAsRegistered(provider);

        if ( this._booted ) {
            await this.bootProvider(provider);
        }

        return provider;
    }

    public markAsRegistered(provider: ServiceProviderContract) {
        this.serviceProviders.push(provider);
    }

    public async boot() {
        if ( this._booted ) {
            return;
        }

        this.serviceProviders.map(async provider => {
            await this.bootProvider(provider);
        });

        this._booted = true;

        await this.fireAppCallbacks(this._bootedCallbacks);
    }

    public async booted(callback: (app: this) => {}) {
        this._bootedCallbacks.push(callback);

        if ( this.isBooted() ) {
            await this.fireAppCallbacks([callback]);
        }
    }

    protected async fireAppCallbacks(callbacks: ((app: this) => {})[]) {
        for (let callback of callbacks) {
            await callback(this);
        }
    }

    public isBooted(): boolean {
        return this._booted;
    }

    public async bootProvider(provider: ServiceProviderContract) {
        if ( typeof (provider as any).boot === 'function' ) {
            await (provider as any).boot();
        }
    }

    public resolveProvider(providerPath: string): ServiceProvider {
        providerPath = this.basePath() ? resolveFrom(this.basePath(), providerPath) : providerPath
        let provider = esmRequire(providerPath);

        const name = path.basename(providerPath, path.extname(providerPath));

        if ( provider[name] ) {
            provider = provider[name] as any;
        }

        if ( typeof (provider) !== 'function' ) {
            throw new Error(`Make sure export default or export ${ path.basename(providerPath) } the provider from ${ providerPath }`)
        }

        return new provider(this);
    }

    public flush() {
        super.flush();

        this.serviceProviders = [];
        this._bootedCallbacks = [];
    }

    protected _environmentPath: string;
    protected _environmentFile: string = '.env';

    public environmentPath() {
        return this._environmentPath || process.cwd();
    }

    public useEnvironmentPath(path: string) {
        this._environmentPath = path;
        return this;
    }

    public loadEnvironmentFrom(file: string) {
        this._environmentFile = file;
    }

    public environmentFile() {
        return this._environmentFile || '.env'
    }

    public environmentFilePath() {
        return path.join(this.environmentPath(), this.environmentFile());
    }

    /**
     * Determine if the application is running unit tests.
     */
    public runningUnitTests(): boolean {
        return this['environment'] === 'test';
    }

    /**
     * Get the path to the cached services.php file.
     *
     * @return string
     */
    public getCachedServicesPath() {
        return this.normalizeCachePath('APP_SERVICES_CACHE', 'cache/services');
    }

    /**
     * Normalize a relative or absolute path to a cache file.
     *
     * @param  string  $key
     * @param  string  $default
     * @return string
     */
    protected normalizeCachePath(key, defaultValue) {
        let env = Env.get(key);
        if ( ! env ) {
            return this.bootstrapPath(defaultValue);
        }

        return env.startsWith('/') ? env : this.basePath(env);
    }

    public bootstrapPath(p = '') {
        return path.join(this.basePath(), 'bootstrap', p);
    }

    /**
     * Get the current application locale.
     *
     * @return string
     */
    public setLocale(locale) {
        this.config.set('i18n.defaultLocale', locale);
        this.translator.setLocale(locale);
        event(new LocaleUpdated(locale), this.events);
    }

    /**
     * Set the current application locale.
     *
     * @param  string  $locale
     * @return void
     */
    public getLocale(): string {
        return this.config.get('i18n.defaultLocale');
    }

    /**
     * Determine if application locale is the given locale.
     *
     * @param  string  $locale
     * @return bool
     */
    public isLocale(locale) {
        return this.getLocale() === locale;
    }
}
