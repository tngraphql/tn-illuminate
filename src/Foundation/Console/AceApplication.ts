/**
 * (c) Phan Trung NguyÃªn <nguyenpl117@gmail.com>
 * User: nguyenpl117
 * Date: 3/14/2020
 * Time: 2:58 PM
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import { Kernel as ConsoleKernel } from '@tngraphql/console';
import { Application } from '../Application';

export class AceApplication extends ConsoleKernel {
    static bootstrappers = [];

    constructor(application?: Application) {
        super(application);

        this.bootstrap();
    }

    async call(command: string, args: any[]): Promise<any> {
        return this.exec(command, args);
    }

    static starting(callback: (ace: AceApplication) => void) {
        this.bootstrappers.push(callback);
    }

    /**
     * Bootstrap the console application.
     *
     */
    protected bootstrap() {
        const bootstrappers = (this.constructor as any).bootstrappers;
        for( const bootstrapper of bootstrappers ) {
            bootstrapper(this);
        }
    }

    /**
     * Clear the console application bootstrappers.
     */
    public static forgetBootstrappers() {
        this.bootstrappers = [];
    }
}
