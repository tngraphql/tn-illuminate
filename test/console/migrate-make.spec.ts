/**
 * (c) Phan Trung Nguyên <nguyenpl117@gmail.com>
 * User: nguyenpl117
 * Date: 3/23/2020
 * Time: 10:12 PM
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

/**
 * (c) Phan Trung Nguyên <nguyenpl117@gmail.com>
 * User: nguyenpl117
 * Date: 3/15/2020
 * Time: 7:41 AM
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

import { Filesystem } from '@poppinss/dev-utils/build';
import { join } from "path";
import { AceApplication, ConsoleKernel, MigrateMakeCommand } from '../../src/Foundation/Console';
import { Application } from '../../src/Foundation';

const fs = new Filesystem(join(__dirname, './migrate'))

describe('Migrate Make', () => {

    afterEach(async () => {
        jest.resetModules();
        await fs.cleanup();
    });

    beforeAll(async () => {
        await fs.add('.env', ``);
        AceApplication.starting(ace => {
            ace.register([MigrateMakeCommand]);
        });
    });

    beforeEach(async () => {
        await fs.add('config/app.ts', `export default {
            providers: []
        }`);
    })

    it('make a new migrate', async () => {
        const app = new Application(fs.basePath);
        app.environment = 'test';

        const kernel: ConsoleKernel = await app.make<ConsoleKernel>(ConsoleKernel);

        const argv = ['make:migration', 'user']
        const command = await kernel.getAce().find(argv)
        const commandInstance = new command!(app)
        await kernel.getAce().runCommand(commandInstance, argv)

        expect(commandInstance.logger.logs[0].startsWith('underline(green(create))')).toBe(true);

        expect(await fs.get(commandInstance.logger.logs[0].replace('underline(green(create)) ', ''))).toMatchSnapshot()
    });
});