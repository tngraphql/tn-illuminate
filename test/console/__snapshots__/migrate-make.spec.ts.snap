// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Migrate Make make a new migrate 1`] = `
"import { Schema as BaseSchema } from '@tngraphql/lucid/build/src/Schema';
import { CreateTableBuilder } from 'knex';

export default class Users extends BaseSchema {
    protected $tableName = 'users'

    public async up () {
        this.schema.createTable(this.$tableName, (table: CreateTableBuilder) => {
            table.increments('id')
            table.timestamps(true)
        })
    }

    public async down () {
        this.schema.dropTable(this.$tableName)
    }
}
"
`;
